### PR TITLE
Support FSDP

### PR DESCRIPTION
**Pull request by tocean**
**Description**
Support FSDP with FP8. 

**Major Revision**
- Add fsdp package
- Add mnist example
- Add FSDPAdam and FSDPAdamW optimizer
- Add document
- Add UT

### CODE REVIEW

**Comment by tocean** [comment]
Log for mnist_fsdp.py:

Train Epoch: 1  Loss: 0.546517
Test set: Average loss: 0.1495, Accuracy: 9549/10000 (95.49%)

Train Epoch: 2  Loss: 0.228595
Test set: Average loss: 0.1186, Accuracy: 9671/10000 (96.71%)

Train Epoch: 3  Loss: 0.155436
Test set: Average loss: 0.0843, Accuracy: 9743/10000 (97.43%)

Train Epoch: 4  Loss: 0.129953
Test set: Average loss: 0.0748, Accuracy: 9780/10000 (97.80%)

Train Epoch: 5  Loss: 0.113996
Test set: Average loss: 0.0673, Accuracy: 9796/10000 (97.96%)

Train Epoch: 6  Loss: 0.104963
Test set: Average loss: 0.0635, Accuracy: 9803/10000 (98.03%)

Train Epoch: 7  Loss: 0.102351
Test set: Average loss: 0.0610, Accuracy: 9813/10000 (98.13%)

Train Epoch: 8  Loss: 0.098512
Test set: Average loss: 0.0599, Accuracy: 9814/10000 (98.14%)

Train Epoch: 9  Loss: 0.096094
Test set: Average loss: 0.0593, Accuracy: 9815/10000 (98.15%)

Train Epoch: 10         Loss: 0.096644
Test set: Average loss: 0.0585, Accuracy: 9820/10000 (98.20%)

Train Epoch: 11         Loss: 0.094363
Test set: Average loss: 0.0580, Accuracy: 9819/10000 (98.19%)

Train Epoch: 12         Loss: 0.094851
Test set: Average loss: 0.0577, Accuracy: 9821/10000 (98.21%)

Train Epoch: 13         Loss: 0.092966
Test set: Average loss: 0.0575, Accuracy: 9819/10000 (98.19%)

Train Epoch: 14         Loss: 0.092716
Test set: Average loss: 0.0574, Accuracy: 9819/10000 (98.19%)

File: msamp/fsdp/_runtime_utils.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+from torch.distributed.fsdp import FullyShardedDataParallel
+
+old_post_backward_hook = torch.distributed.fsdp._runtime_utils._post_backward_hook
+
+
+@no_type_check
+@torch.no_grad()
+def _fp8_post_backward_hook(state, handle, *unused):
+    """A post-backward communication hook which supports fp8."""
+    if not isinstance(state, FullyShardedDataParallel):
```
Comment (wkcn) [inline msamp/fsdp/_runtime_utils.py:18]: Why is the old_post_backward_hook called when the state is not FullyShardedDataParallel?

**Review by wkcn — COMMENTED** [review]
Good job!

I have the following questions:
1. Which line of the code synchroinize the scaling factor (meta.scale or meta.scale_inv) across GPUs?
2. Gradient accumulation seems to be not supported yet. We can raise a NotImplemented exception when the steps of gradient accumulation is larger than 1.
3. Is there any comparision on memory footprint between BF16-FSDP and FP8-FSDP?

**Comment by tocean** [comment]
> 1. Which line of the code synchroinize the scaling factor (meta.scale or meta.scale_inv) across GPUs?
> 2. Gradient accumulation seems to be not supported yet. We can raise a NotImplemented exception when the steps of gradient accumulation is larger than 1.
> 3. Is there any comparision on memory footprint between BF16-FSDP and FP8-FSDP?

For these 3 questions:
1. See `FSDPAdamW::step`. In this function, all_reduce of amax is called.
2. Add check loigc in _fp8_post_backward_hook. 
3. I checked the memory saving use T5 example [here](https://github.com/pytorch/examples/blob/main/distributed/FSDP/T5_training.py). The model I use is t5-3b. The memory footprint for bf16, FP32, MS-AMP is 28GB, 40GB,34GB. It is a bit strange why MS-AMP uses more memory than BF16.

**Comment by wkcn** [comment]
> > 1. Which line of the code synchroinize the scaling factor (meta.scale or meta.scale_inv) across GPUs?
> > 2. Gradient accumulation seems to be not supported yet. We can raise a NotImplemented exception when the steps of gradient accumulation is larger than 1.
> > 3. Is there any comparision on memory footprint between BF16-FSDP and FP8-FSDP?
> 
> For these 3 questions:
> 
> 1. See `FSDPAdamW::step`. In this function, all_reduce of amax is called.
> 2. Add checking loigc in _fp8_post_backward_hook.
> 3. I checked the memory saving use T5 example [here](https://github.com/pytorch/examples/blob/main/distributed/FSDP/T5_training.py). The model I use is t5-3b. The memory footprint for bf16, FP32, MS-AMP is 28GB, 40GB and 34GB respectively. It is a bit strange that MS-AMP uses more memory than BF16.

Regarding to the answer 3, it may be related to the argument [`mixed_precision` of FSDP](https://github.com/pytorch/examples/blob/main/distributed/FSDP/T5_training.py#L119). 

When `mixed_precision` is not None, FSDP will create a FP32 master weight for FP8Linear. It leads to duplicated master weights in FSDP and FP8Optimizer.

File: msamp/fsdp/flat_param.py [hunk 2]
```diff
@@ -0,0 +1,71 @@
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+"""MS-AMP fsdp.flat_param module."""
+
+from typing import Optional, Sequence
+
+import torch
+import torch.nn as nn
+import torch.distributed as dist
```
Comment (wkcn) [inline msamp/fsdp/flat_param.py:10]: `dist` is unused.

Comment (tocean) [reply msamp/fsdp/flat_param.py]: Fixed.

File: msamp/fsdp/_runtime_utils.py (hunk 1 above)
Comment (tocean) [reply msamp/fsdp/_runtime_utils.py]: Good question. Seems state is always instance of FullyShardedDataParallel, so I remove the check `if not isinstance(state, FullyShardedDataParallel)`.

File: msamp/fsdp/_runtime_utils.py [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+old_post_backward_hook = torch.distributed.fsdp._runtime_utils._post_backward_hook
+
+
+@no_type_check
+@torch.no_grad()
+def _fp8_post_backward_hook(state, handle, *unused):
+    """A post-backward communication hook which supports fp8."""
+    accumulate_grad = hasattr(state._flat_param, '_saved_grad_shard')
+    if accumulate_grad and not torch.all(state._flat_param._saved_grad_shard == 0):
```
Comment (wkcn) [inline msamp/fsdp/_runtime_utils.py:18]: The following statement is faster.
`if accumulate_grad and not torch.count_nonzero(state._flat_param._saved_grad_shard) == 0:`

File: msamp/fsdp/_runtime_utils.py [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+        raise NotImplementedError('accumulate_grad is not supported for fp8')
```
Comment (wkcn) [inline msamp/fsdp/_runtime_utils.py:19]: accumulate_grad is not supported yet for fp8'

File: msamp/optim/adamw.py [hunk 5]
```diff
@@ -228,3 +229,105 @@ def adamw_fn(   # noqa: C901
...
+        for i, param in enumerate(self.original_params):
+            if hasattr(param, '_meta') and param._meta is not None:
+                hp_data = None
+                if param.numel() == 0:
+                    param._meta.amax[0].zero_()
+                else:
+                    hp_data = self.master_weights[i].float()
+                    param._meta.amax[0] = hp_data.abs().max()
+
+                dist.all_reduce(param._meta.amax[0], op=dist.ReduceOp.MAX)
```
Comment (wkcn) [inline msamp/optim/adamw.py:325]: Why synchronize amax here?

`param` is shard in FSDP and stored only in a specific GPU.

I think it is unnecessary to synchronize `amax`.

Comment (tocean) [reply msamp/optim/adamw.py:325]: Just as synced offline, it is possible that one parameter is sharded to multiple GPUs, so it is necessary to synchronize `amax`.

File: msamp/fsdp/_runtime_utils.py (hunk 3 above)
Comment (tocean) [reply msamp/fsdp/_runtime_utils.py]: Fixed.

File: msamp/fsdp/_runtime_utils.py (hunk 4 above)
Comment (tocean) [reply msamp/fsdp/_runtime_utils.py]: Fixed.

**Review by wkcn — APPROVED** [review]
LGTM. Thanks!

**Review by penghouwen — APPROVED** [review]
an initial version of FSDP in MSAMP

**Comment by tocean** [comment]
> In the future, we will do:
> 
> * gradient accumulation
> * speed optimization
> * acc calibration

Sure. Let's improve it in next iteration.
